mikrotik/resource_bgp_instance.go

```diff
@@ -134,102 +137,77 @@ func resourceBgpInstanceUpdate(d *schema.ResourceData, m interface{}) error {
+	c := m.(*client.Mikrotik)
 
 	err := c.DeleteBgpInstance(d.Get("name").(string))
 
 	if err != nil {
-		return err
+		return diag.FromErr(err)
 	}
 
 	d.SetId("")
 	return nil
 }
 
-func bgpInstanceToData(b *client.BgpInstance, d *schema.ResourceData) error {
+func bgpInstanceToData(b *client.BgpInstance, d *schema.ResourceData) diag.Diagnostics {
+	values := map[string]interface{}{
+		"name":                        b.Name,
+		"as":                          b.As,
+		"client_to_client_reflection": b.ClientToClientReflection,
+		"comment":                     b.Comment,
+		"confederation_peers":         b.ConfederationPeers,
+		"disabled":                    b.Disabled,
+		"ignore_as_path_len":          b.IgnoreAsPathLen,
+		"out_filter":                  b.OutFilter,
+		"redistribute_connected":      b.RedistributeConnected,
+		"redistribute_ospf":           b.RedistributeOspf,
+		"redistribute_other_bgp":      b.RedistributeOtherBgp,
+		"redistribute_rip":            b.RedistributeRip,
+		"redistribute_static":         b.RedistributeStatic,
+		"router_id":                   b.RouterID,
+		"routing_table":               b.RoutingTable,
+		"cluster_id":                  b.ClusterID,
+		"confederation":               b.Confederation,
+	}
+
 	d.SetId(b.Name)
 
-	if err := d.Set("name", b.Name); err != nil {
-		return err
-	}
-	if err := d.Set("as", b.As); err != nil {
-		return err
-	}
-	if err := d.Set("client_to_client_reflection", b.ClientToClientReflection); err != nil {
-		return err
-	}
-	if err := d.Set("comment", b.Comment); err != nil {
-		return err
-	}
-	if err := d.Set("confederation_peers", b.ConfederationPeers); err != nil {
-		return err
-	}
-	if err := d.Set("disabled", b.Disabled); err != nil {
-		return err
-	}
-	if err := d.Set("ignore_as_path_len", b.IgnoreAsPathLen); err != nil {
-		return err
-	}
-	if err := d.Set("out_filter", b.OutFilter); err != nil {
-		return err
-	}
-	if err := d.Set("redistribute_connected", b.RedistributeConnected); err != nil {
-		return err
-	}
-	if err := d.Set("redistribute_ospf", b.RedistributeOspf); err != nil {
-		return err
-	}
-	if err := d.Set("redistribute_other_bgp", b.RedistributeOtherBgp); err != nil {
-		return err
-	}
-	if err := d.Set("redistribute_rip", b.RedistributeRip); err != nil {
-		return err
-	}
-	if err := d.Set("redistribute_static", b.RedistributeStatic); err != nil {
-		return err
-	}
-	if err := d.Set("router_id", b.RouterID); err != nil {
-		return err
-	}
-	if err := d.Set("routing_table", b.RoutingTable); err != nil {
-		return err
-	}
-	if err := d.Set("cluster_id", b.ClusterID); err != nil {
-		return err
-	}
-	if err := d.Set("confederation", b.Confederation); err != nil {
-		return err
+	var diags diag.Diagnostics
+
+	for key, value := range values {
+		if err := d.Set(key, value); err != nil {
+			diags = append(diags, diag.Errorf("failed to set %s: %v", key, err)...)
+		}
 	}
```

Comment:
This is so much nicer. We could probably make this even better with a function like so since we have this same loop multiple times.

```
func SetResourceDataAttrs(d *schema.ResourceData, attrs map[string]interface{}) diags diag.Diagnostics {
  var diags diag.Diagnostics
  for key, value := range values {
		if err := d.Set(key, value); err != nil {
			diags = append(diags, diag.Errorf("failed to set %s: %v", key, err)...)
		}
	}
  return diags
}
```

This PR is getting big so I don't think we need to do it now, but we should follow up with that afterwards.